File: python/liquid_crystal_virtual.py
#!/usr/bin/python
# coding: UTF-8

import queue
import threading
import time
from lcd_core.liquid_crystal_mock_tk import LiquidCrystalMockTk

class LiquidCrystalVirtual:
    def __init__(self, lcd, size_x=16, size_y=2, blank_ch=" "):
        self.lcd = lcd
        self._max_x = size_x
        self._max_y = size_y
        self._blank_ch = blank_ch
        self._vram = [[self._blank_ch] * self._max_x for _ in range(self._max_y)]
        self._is_dirty = False
    @property
    def max_x(self):
        return self._max_x
    @property
    def max_y(self):
        return self._max_y
    def clear(self):
        for y in range(self._max_y):
            for x in range(self._max_x):
                self._vram[y][x] = self._blank_ch
        self._is_dirty = True

    def print(self, text, x, y):
        if y < 0 or y >= self._max_y:
            return
        for i, ch in enumerate(str(text)):
            tx = x + i
            if 0 <= tx < self._max_x:
                if self._vram[y][tx] != ch:
                    self._vram[y][tx] = ch
                    self._is_dirty = True

    def print_clipped(self, text, x, y, clip_left, clip_right):
        """指定した列の範囲（clip_left ～ clip_right）内だけに文字を描画する（マスク処理）"""
        # 入力されたクリップ範囲をVRAMの有効幅に収める
        w_left = max(0, clip_left)
        w_right = min(self._max_x - 1, clip_right)
        
        if y < 0 or y >= self._max_y:
            return

        for i, ch in enumerate(str(text)):
            tx = x + i
            # クリップ範囲の内側、かつVRAMの範囲内にある文字だけを書き込む
            if w_left <= tx <= w_right:
                if self._vram[y][tx] != ch:
                    self._vram[y][tx] = ch
                    self._is_dirty = True

    def flush(self):
        if not self._is_dirty: return
        for y in range(self._max_y):
            self.lcd.set_cursor(0, y)
            self.lcd.print("".join(self._vram[y]))
        self._is_dirty = False


def loop(data_queue, vlcd):
    while True:
        vlcd.flush()

        data_queue.put("update")
        time.sleep(0.05)

if __name__ == "__main__":
    shared_queue = queue.Queue()
    size_x = 16
    size_y = 4
    lcd = LiquidCrystalMockTk(shared_queue, size_x, size_y, blank_ch="_")
    vlcd = LiquidCrystalVirtual(lcd, size_x=size_x, size_y=size_y)
    vlcd.clear()
    vlcd.print_clipped("Hello, World!", 0, 0, 2, 4)
    vlcd.print("Blinking Text", 0, 1)
    vlcd.print("Sliding Text", 2, 2)
    vlcd.print("Knight Rider", 3, 3)
    loop_thread = threading.Thread(
        target=loop, args=(shared_queue, vlcd, ), daemon=True
    )
    loop_thread.start()

    lcd.mainloop()

File: cpp/LiquidCrystalVirtual.cpp
#include "LiquidCrystalVirtual.h"

// コンストラクタ
LiquidCrystalVirtual::LiquidCrystalVirtual(ILcdDevice* lcd, uint8_t size_x, uint8_t size_y, char blank_ch) {
    _lcd = lcd;
    _max_x = size_x;
    _max_y = size_y;
    _blank_ch = blank_ch;
    _is_dirty = false;

    _vram = new char*[_max_y];
    for (uint8_t y = 0; y < _max_y; ++y) {
        _vram[y] = new char[_max_x + 1];
        for (uint8_t x = 0; x < _max_x; ++x) {
            _vram[y][x] = _blank_ch;
        }
        _vram[y][_max_x] = '\0';
    }
}

// デストラクタ
LiquidCrystalVirtual::~LiquidCrystalVirtual() {
    for (uint8_t y = 0; y < _max_y; ++y) {
        delete[] _vram[y];
    }
    delete[] _vram;
}

uint8_t LiquidCrystalVirtual::max_x() const { return _max_x; }
uint8_t LiquidCrystalVirtual::max_y() const { return _max_y; }
bool LiquidCrystalVirtual::is_dirty() const { return _is_dirty; }

void LiquidCrystalVirtual::clear() {
    for (uint8_t y = 0; y < _max_y; ++y) {
        for (uint8_t x = 0; x < _max_x; ++x) {
            _vram[y][x] = _blank_ch;
        }
    }
    _is_dirty = true;
}

void LiquidCrystalVirtual::print(const char* text, int16_t x, int16_t y) {
    if (y < 0 || y >= _max_y) return;

    int16_t i = 0;
    while (text[i] != '\0') {
        int16_t tx = x + i;
        if (tx >= 0 && tx < _max_x) {
            if (_vram[y][tx] != text[i]) {
                _vram[y][tx] = text[i];
                _is_dirty = true;
            }
        }
        i++;
    }
}

void LiquidCrystalVirtual::print_clipped(const char* text, int16_t x, int16_t y, int16_t clip_left, int16_t clip_right) {
    if (y < 0 || y >= _max_y) return;

    int16_t w_left = (clip_left > 0) ? clip_left : 0;
    int16_t w_right = (clip_right < _max_x - 1) ? clip_right : _max_x - 1;

    int16_t i = 0;
    while (text[i] != '\0') {
        int16_t tx = x + i;
        if (tx >= w_left && tx <= w_right) {
            if (_vram[y][tx] != text[i]) {
                _vram[y][tx] = text[i];
                _is_dirty = true;
            }
        }
        i++;
    }
}

void LiquidCrystalVirtual::flush() {
    if (!_is_dirty) return;
    if (_lcd == nullptr) return;

    for (uint8_t y = 0; y < _max_y; ++y) {
        _lcd->set_cursor(0, y);
        _lcd->print(_vram[y]);
    }
    _is_dirty = false;
}











